File: src/page/info/HeadTable.java
package page.info;

import java.awt.Point;
import java.text.DecimalFormat;

import common.util.stage.Castles;
import common.util.stage.CharaGroup;
import common.util.stage.Limit;
import common.util.stage.LvRestrict;
import common.util.stage.Stage;
import page.MainFrame;
import page.MainLocale;
import page.Page;
import page.pack.CharaGroupPage;
import page.pack.LvRestrictPage;
import page.support.AbJTable;
import page.view.BGViewPage;
import page.view.CastleViewPage;
import page.view.MusicPage;

public class HeadTable extends AbJTable {

	private static final long serialVersionUID = 1L;

	private static String[] infs, limits, rarity;

	static {
		redefine();
	}

	public static void redefine() {
		infs = Page.get(1, "ht0", 6);
		limits = Page.get(1, "ht1", 7);
		rarity = new String[] { "N", "EX", "R", "SR", "UR", "LR" };
	}

	private Object[][] data;
	private Stage sta;
	private final Page page;

	protected HeadTable(Page p) {
		page = p;
	}

	@Override
	public Class<?> getColumnClass(int c) {
		return String.class;
	}

	@Override
	public int getColumnCount() {
		return 8;
	}

	@Override
	public String getColumnName(int arg0) {
		return "";
	}

	@Override
	public int getRowCount() {
		if (data == null)
			return 0;
		return data.length;
	}

	@Override
	public Object getValueAt(int r, int c) {
		if (data == null || r < 0 || c < 0 || r >= data.length || c >= data[r].length)
			return null;
		return data[r][c];
	}

	protected void clicked(Point p) {
		if (data == null)
			return;
		int c = getColumnModel().getColumnIndexAtX(p.x);
		int r = p.y / getRowHeight();
		if (r == 1 && c == 5)
			MainFrame.changePanel(new MusicPage(page, sta.mus0));
		if (r == 1 && c == 7)
			MainFrame.changePanel(new MusicPage(page, sta.mus1));
		if (r == 2 && c == 1)
			MainFrame.changePanel(new BGViewPage(page, null, sta.bg));
		if (r == 2 && c == 3)
			MainFrame.changePanel(new CastleViewPage(page, Castles.map.values(), sta.getCastle()));
		if (r == 2 && c == 5 && data[r][c] != null && data[r][c] instanceof LvRestrict)
			MainFrame.changePanel(new LvRestrictPage(page, (LvRestrict) data[r][c]));
		if (r == 2 && c == 7 && data[r][c] != null)
			MainFrame.changePanel(new CharaGroupPage(page, (CharaGroup) data[r][c]));
	}

	protected void setData(Stage st) {
		sta = st;
		Object[][] lstr = new Object[5][8];
		Object[] tit, bas, img, rar, reg;
		tit = lstr[0];
		bas = lstr[1];
		img = lstr[2];
		rar = lstr[3];
		reg = lstr[4];
		tit[0] = "ID:";
		tit[1] = st.map.mc.name + "-" + st.map.id + "-" + st.id();
		String star = Page.get(1, "star");
		for (int i = 0; i < st.map.stars.length; i++)
			tit[2 + i] = (i + 1) + star + ": " + st.map.stars[i] + "%";
		bas[0] = infs[0];
		bas[1] = st.health;
		bas[2] = infs[1] + ": " + st.len;
		bas[3] = infs[2] + ": " + st.max;
		bas[4] = Page.get(1, "mus") + ":";
		bas[5] = st.mus0;
		bas[6] = "<" + st.mush + "%:";
		bas[7] = st.mus1;
		img[0] = infs[4];
		img[1] = st.bg;
		img[2] = infs[5];
		img[3] = st.getCastle();
		img[4] = MainLocale.getLoc(1, "lop");
		img[5] = convertTime(st.loop0);
		img[6] = MainLocale.getLoc(1, "lop1");
		img[7] = convertTime(st.loop1);
		Limit lim = st.getLim(0);
		if (lim != null) {
			if (lim.rare != 0) {
				rar[0] = limits[0];
				int j = 1;
				for (int i = 0; i < rarity.length; i++)
					if (((lim.rare >> i) & 1) > 0)
						rar[j++] = rarity[i];
			}
			if (lim.lvr != null) {
				img[4] = limits[6];
				img[5] = lim.lvr;
			}
			if (lim.group != null) {
				img[6] = limits[5];
				img[7] = lim.group;
			}
			if (lim.min + lim.max + lim.max + lim.line + lim.num > 0) {
				int i = 0;
				if (lim.min > 0) {
					reg[0] = limits[3];
					reg[1] = "" + lim.min;
					i = 2;
				}
				if (lim.max > 0) {
					reg[i] = limits[4];
					reg[i + 1] = "" + lim.max;
					i += 2;
				}
				if (lim.num > 0) {
					reg[i] = limits[1];
					reg[i + 1] = "" + lim.num;
					i += 2;
				}
				if (lim.line > 0)
					reg[i] = limits[2];
			}
		}
		data = lstr;
	}

	private String convertTime(long milli) {
		long min = milli / 60 / 1000;

		double time = milli - (double) min * 60000;

		time /= 1000;

		DecimalFormat df = new DecimalFormat("#.###");

		double s = Double.parseDouble(df.format(time));

		if (s >= 60) {
			s -= 60;
			min += 1;
		}

		if (s < 10) {
			return min + ":" + "0" + df.format(s);
		} else {
			return min + ":" + df.format(s);
		}
	}

}

File: src/page/info/HeadTable.kt
package page.info

import common.util.stage.CastleList
import common.util.stage.CharaGroup
import common.util.stage.LvRestrict
import common.util.stage.Stage
import page.MainFrame
import page.MainLocale
import page.Page
import page.info.HeadTable
import page.pack.CharaGroupPage
import page.pack.LvRestrictPage
import page.support.AbJTable
import page.view.BGViewPage
import page.view.CastleViewPage
import page.view.MusicPage
import java.awt.Point
import java.text.DecimalFormat

class HeadTable(private val page: Page) : AbJTable() {
    companion object {
        private const val serialVersionUID = 1L
        private val infs: Array<String>
        private val limits: Array<String>
        private val rarity: Array<String>
        fun redefine() {
            HeadTable.Companion.infs = Page.Companion.get(1, "ht0", 6)
            HeadTable.Companion.limits = Page.Companion.get(1, "ht1", 7)
            HeadTable.Companion.rarity = arrayOf("N", "EX", "R", "SR", "UR", "LR")
        }

        init {
            HeadTable.Companion.redefine()
        }
    }

    private var data: Array<Array<Any>>?
    private var sta: Stage? = null
    override fun getColumnClass(c: Int): Class<*> {
        return String::class.java
    }

    override fun getColumnCount(): Int {
        return 8
    }

    override fun getColumnName(arg0: Int): String {
        return ""
    }

    override fun getRowCount(): Int {
        return if (data == null) 0 else data!!.size
    }

    override fun getValueAt(r: Int, c: Int): Any {
        return if (data == null || r < 0 || c < 0 || r >= data!!.size || c >= data!![r].length) null else data!![r][c]
    }

    fun clicked(p: Point) {
        if (data == null) return
        val c: Int = getColumnModel().getColumnIndexAtX(p.x)
        val r: Int = p.y / getRowHeight()
        if (r == 1 && c == 5) MainFrame.Companion.changePanel(MusicPage(page, sta!!.mus0))
        if (r == 1 && c == 7) MainFrame.Companion.changePanel(MusicPage(page, sta!!.mus1))
        if (r == 2 && c == 1) MainFrame.Companion.changePanel(BGViewPage(page, null, sta!!.bg))
        if (r == 2 && c == 3) MainFrame.Companion.changePanel(CastleViewPage(page, CastleList.Companion.from(sta), sta!!.castle))
        if (r == 2 && c == 5 && data!![r][c] != null && data!![r][c] is LvRestrict) MainFrame.Companion.changePanel(LvRestrictPage(page, data!![r][c] as LvRestrict))
        if (r == 2 && c == 7 && data!![r][c] != null) MainFrame.Companion.changePanel(CharaGroupPage(page, data!![r][c] as CharaGroup))
    }

    fun setData(st: Stage) {
        sta = st
        val lstr = Array<Array<Any>>(5) { arrayOfNulls(8) }
        val tit: Array<Any>
        val bas: Array<Any>
        val img: Array<Any>
        val rar: Array<Any>
        val reg: Array<Any>
        tit = lstr[0]
        bas = lstr[1]
        img = lstr[2]
        rar = lstr[3]
        reg = lstr[4]
        tit[0] = "ID:"
        tit[1] = st.map.mc.toString() + "-" + st.map.id + "-" + st.id()
        val star: String = page.Page.Companion.get(1, "star")
        for (i in st.map.stars.indices) tit[2 + i] = (i + 1).toString() + star + ": " + st.map.stars[i] + "%"
        bas[0] = page.info.HeadTable.Companion.infs.get(0)
        bas[1] = st.health
        bas[2] = page.info.HeadTable.Companion.infs.get(1).toString() + ": " + st.len
        bas[3] = page.info.HeadTable.Companion.infs.get(2).toString() + ": " + st.max
        bas[4] = page.Page.Companion.get(1, "mus") + ":"
        bas[5] = st.mus0
        bas[6] = "<" + st.mush + "%:"
        bas[7] = st.mus1
        img[0] = page.info.HeadTable.Companion.infs.get(4)
        img[1] = st.bg
        img[2] = page.info.HeadTable.Companion.infs.get(5)
        img[3] = st.castle
        img[4] = MainLocale.Companion.getLoc(1, "lop")
        img[5] = convertTime(st.loop0)
        img[6] = MainLocale.Companion.getLoc(1, "lop1")
        img[7] = convertTime(st.loop1)
        val lim = st.getLim(0)
        if (lim != null) {
            if (lim.rare != 0) {
                rar[0] = page.info.HeadTable.Companion.limits.get(0)
                var j = 1
                for (i in page.info.HeadTable.Companion.rarity.indices) if (lim.rare shr i and 1 > 0) rar[j++] = page.info.HeadTable.Companion.rarity.get(i)
            }
            if (lim.lvr != null) {
                img[4] = page.info.HeadTable.Companion.limits.get(6)
                img[5] = lim.lvr
            }
            if (lim.group != null) {
                img[6] = page.info.HeadTable.Companion.limits.get(5)
                img[7] = lim.group
            }
            if (lim.min + lim.max + lim.max + lim.line + lim.num > 0) {
                var i = 0
                if (lim.min > 0) {
                    reg[0] = page.info.HeadTable.Companion.limits.get(3)
                    reg[1] = "" + lim.min
                    i = 2
                }
                if (lim.max > 0) {
                    reg[i] = page.info.HeadTable.Companion.limits.get(4)
                    reg[i + 1] = "" + lim.max
                    i += 2
                }
                if (lim.num > 0) {
                    reg[i] = page.info.HeadTable.Companion.limits.get(1)
                    reg[i + 1] = "" + lim.num
                    i += 2
                }
                if (lim.line > 0) reg[i] = page.info.HeadTable.Companion.limits.get(2)
            }
        }
        data = lstr
    }

    private fun convertTime(milli: Long): String {
        var min = milli / 60 / 1000
        var time = milli - min.toDouble() * 60000
        time /= 1000.0
        val df = DecimalFormat("#.###")
        var s: Double = df.format(time).toDouble()
        if (s >= 60) {
            s -= 60.0
            min += 1
        }
        return if (s < 10) {
            min.toString() + ":" + "0" + df.format(s)
        } else {
            min.toString() + ":" + df.format(s)
        }
    }
}
